src/include/migraphx/register_op.hpp

```diff
@@ -33,15 +33,33 @@
 namespace migraphx {
 inline namespace MIGRAPHX_INLINE_NS {
 
+// unregister all ops for specified target, useful when unloading dynamically plugged-in target lib
+void unregister_op(const std::string& op_name);
+
+struct op_handler
+{
+    operation op;
+    std::string name;
+    op_handler(const operation& op_r) : op(op_r), name(op.name()){};
+    ~op_handler() { unregister_op(name); }
+};
+
+std::unordered_map<std::string, operation>& op_map();
```

Comment:
The map should not be exposed.

Reply:
Also, the map needs to be changed to `std::unordered_map<std::string, const operation*>`.

Reply:
That map is exposed because it needs to be instantiated first as `static` object.  `static` objects gets destructed in the order they are defined.  I would need to instantiate `static` map before registering any `static` operator, during `register_op`.

Reply:
If we make `operation*` it would require more changes in `load_op` and `make_op` etc. 

Why does it need to be a `operation*` ? with the current changes it seems alright.

Reply:
> I would need to instantiate static map before registering any static operator, during register_op.

`register_op(op_h.op)` already calls the static map so it should be created already. I dont see why it needs to be in `register_op<T>()`.

> Why does it need to be a operation* ?

So the memory will be owned by target .so, rather than libmigraphx.so.

> with the current changes it seems alright.

Yea it may not be an issue since we always unregister it.

Reply:
> register_op(op_h.op) already calls the static map so it should be created already. I dont see why it needs to be in register_op<T>().

During unloading I found sometimes otherwise it destructs static `op_map` first and then tries to call `unregister_op()`which requires `op_map()` to be present.   
Therefore calling `op_map()` first inside `register_op<T>()` gurantees destruction order.

Reply:
> it destructs static op_map first and then tries to call unregister_op()which requires op_map() to be present.

So does it construct the `op_map` again? Or is there an error about missing the operator in the map?

> Therefore calling op_map() first inside register_op<T>() gurantees destruction order.

We dont want to expose these global variables in the header.

Reply:
I see the issue, the `op_map` needs to be called before the static variable, but I think we can encapsulate that in a function.

Reply:
Yea we can add a `register_op_init` function we call that can then calls `op_map` to construct it.

Reply:
Added `op_init()`